app/src/main/java/unipd/se18/ocrcamera/PreProcessing.java

```diff
@@ -0,0 +1,369 @@
+
+        //Sums the matrix of the colors into a single one
+        Mat lumTemp = new Mat();
+        Mat lum = new Mat();
+        Core.add(lumRed , lumGreen , lumTemp); //lumRed + lumGreen = lumTemp
+        Core.add(lumTemp , lumBlue , lum); //lumBlue + lumTemp = lum
+
+        //Calculate the sum of the values of all pixels
+        Scalar sum = Core.sumElems(lum);
+
+        //Image's bit
+        int bit;
+        switch ( brightnessMat.depth() ) {
+            case CV_8U:  bit = 8; break;
+            case CV_8S:  bit = 8; break;
+            case CV_16U: bit = 16; break;
+            case CV_16S: bit = 16; break;
+            case CV_32S: bit = 32; break;
+            case CV_32F: bit = 32; break;
+            case CV_64F: bit = 64; break;
+            default: return 0;
+        }
+        //Calculate the percentage of the brightness
+        double brightness = sum.val[0]/((Math.pow(2,bit)-1)*brightnessMat.rows()*brightnessMat.cols())*2;
+
+        Log.d(TAG, "Brightness:"+brightness);
+
+        /*Bounds to define if the image is dark or bright.
+          The values were decided on the basis of various tests*/
+        double upperBound = 0.9;
+        double lowerBound = 0.4;
+
+        if (brightness > upperBound){             //Image is too bright
+            return 1;
+        } else if (brightness < lowerBound){        //Image is too dark
+            return 2;
+        } else {      //Image is neither too bright nor too dark
+            return 0;
+        }
+    }
+
+
+    /**
+     * Change the brightness of the image into an optimal one
+     * @param image the image we want to modify the brightness
+     * @return the image with the modified brightness
+     * @author Thomas Porro(g1), Giovanni Fasan(g1), Oscar Garrido(g1)
+     */
+    private Bitmap editBright(Bitmap image){
+        //Converts the image into a matrix
+        Mat bright;
+        try{
+            bright = IPUtils.conversionBitmapToMat(image);
+        } catch (ConversionFailedException error){
+            Log.e(TAG, error.getErrorMessage());
+            return image;
+        }
+
+        //Call the internal method isBright to detect if the image is bright or dark
+        //and change the brightness according to the number obtained
+        Mat modifiedMat = new Mat();
+        switch (isBright(bright)) {
+            case 1: //If the image is too bright
+                Log.d(TAG, "Case==1 ==> Too bright");
+                //Darkens the colour's brightness until it's in an optimal value
+                for(double changeBrightness=0; changeBrightness!=-240; changeBrightness-=15){
+                    //Converts an array to another data type with optional scaling.
+                    bright.convertTo(modifiedMat, -1, 1, changeBrightness);
+                    if(isBright(modifiedMat)==0){
+                        try{
+                            return IPUtils.conversionMatToBitmap(modifiedMat);
```

Comment:
I think that you shoudn't put a return statement in a for. it's generally better to use one return statement because it is easier to figure out where the method will exit.

Reply:
you're probably right. I will fix the code to make the return out of the for